### PR TITLE
Added stats screens

### PR DESCRIPTION
**Pull request by MaticTomislav**
### Pull request description

- What is new?
  * Implemented statistics screens 
  * basic silos algorithm
  
- What has changed?
  * Changed

### Checklist
- [x] I ran this code locally

### CODE REVIEW

File: services/rule_engine/src/main/resources/rules/Ava/Statistics.drl [hunk 1]
```diff
@@ -26,5 +30,26 @@ rule "Send stats email"
...
+        agent.send(CultureMaster.ID, new NumOfSilosSignal(agent.getId()));
         agent.sendStatisticsEmail(employee);
+end
+
+rule "Store num of silos"
+    when
+        signal: SilosInfoSignal() from entry-point "signals"
+        agent: Ava()
+    then
+        Log.info("primio num of silos:" + signal.getNumOfSilos());
```
Comment (mmatichr) [inline services/rule_engine/src/main/resources/rules/Ava/Statistics.drl]: delete

File: services/rule_engine/src/main/resources/rules/CultureMaster/CultureMaster.drl [hunk 2]
```diff
@@ -45,3 +47,12 @@ rule "Trigger familiarity quiz"
         Log.info("CultureMaster: triggered familiarity quiz from dashboard");
         // trigger quiz
 end
+
+rule "Juraj smisli ime"
```
Comment (mmatichr) [inline services/rule_engine/src/main/resources/rules/CultureMaster/CultureMaster.drl]: smisli

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 3]
```diff
@@ -79,4 +79,95 @@ public void sendEmployeesToAva() {
...
+            for(Map.Entry<String, Double> entry : employee.getFamiliarity().entrySet()){
+                if(j==i) {
+                    matrix[i][j] = 0.0;
+                    binaryMatrix[i][j]=0;
+                    j++;
+                }
+
+                matrix[i][j] = entry.getValue();
+                if( matrix[i][j]>limit)
+                binaryMatrix[i][j]=1;
```
Comment (mmatichr) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: formatiraj

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 4]
```diff
@@ -103,9 +104,9 @@ public int[][] calculateScore(Map<String, EmployeeProfile> employees){
 
                 matrix[i][j] = entry.getValue();
                 if( matrix[i][j]>limit)
```
Comment (mmatichr) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: whitespace

File: services/mitems/data.json [hunk 5]
```diff
@@ -203,6 +203,31 @@
             "content": "Statistics"
           }
         ]
+      },
+      {
+        "name": "Stats screens",
```
Comment (mjuraj) [inline services/mitems/data.json:263]: Split into separate flow and separate screens

File: services/rule_engine/src/main/java/agents/Ava.java [hunk 6]
```diff
@@ -39,6 +39,7 @@ public class Ava extends Agent {
     private Map<String, EmployeeProfile> otherEmployees;
     private boolean workingHours;
     private Date statsEmailLastSentAt;
+    private int numOfSilos;
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/Ava.java]: numOf -> NOOOOOOOOOO

Comment (mjuraj) [reply services/rule_engine/src/main/java/agents/Ava.java]: silosCount

File: services/rule_engine/src/main/java/agents/Ava.java [hunk 7]
```diff
@@ -255,4 +256,38 @@ public void sendStatisticsEmail(EmployeeProfile employee) throws IOException {
         e.setHtmlText(htmlBody);
         EmailAdapterAPI.newEmail(e);
     }
+
+    public void showEmployeeNumberScreen() {
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/Ava.java]: Not just a single screen

File: services/rule_engine/src/main/java/agents/Ava.java [hunk 8]
```diff
@@ -255,4 +256,38 @@ public void sendStatisticsEmail(EmployeeProfile employee) throws IOException {
...
+        Map<String, BaseTemplate> screens = new HashMap<String, BaseTemplate>();
+        //String nextButton = ;
+        String employeeNumberScreenDescription = Mitems.getText("statistics.stats-screens.employee-number-screen-description");
+        String employeeNumberScreenNumber = String.format("%d",otherEmployees.values().size()+1);
+
+        screens.put("employeeNumberScreen", new TemplateGenerator()
+                .addComponent("title", new DescriptionComponent(employeeNumberScreenDescription))
+                .addComponent("description", new TitleComponent(employeeNumberScreenNumber))
+                .addComponent("submit", new PrimarySubmitButtonComponent("Next", "silosNumberScreen")));
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/Ava.java]: Text of each button needs to be on mitems

File: services/rule_engine/src/main/java/agents/Ava.java [hunk 9]
```diff
@@ -255,4 +256,38 @@ public void sendStatisticsEmail(EmployeeProfile employee) throws IOException {
...
         EmailAdapterAPI.newEmail(e);
     }
+
+    public void showEmployeeNumberScreen() {
+        Map<String, BaseTemplate> screens = new HashMap<String, BaseTemplate>();
+        //String nextButton = ;
+        String employeeNumberScreenDescription = Mitems.getText("statistics.stats-screens.employee-number-screen-description");
+        String employeeNumberScreenNumber = String.format("%d",otherEmployees.values().size()+1);
+
+        screens.put("employeeNumberScreen", new TemplateGenerator()
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/Ava.java:300]: Add back buttons

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 10]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
         }
     }
 
+    public int[][] calculateScore(Map<String, EmployeeProfile> employees){
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: What score

Comment (mjuraj) [reply services/rule_engine/src/main/java/agents/CultureMaster.java]: function name doesn't match the return value

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 11]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+
+        Double[][] matrix = new Double[employees.values().size()][employees.values().size()]; //weight matrix
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: `//<space>text`

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 12]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+        int[][] binaryMatrix = new int[employees.values().size()][employees.values().size()]; // binary matrix
+        
+        List<String> list = new LinkedList<>();
+        int limit=3;
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: limit = 3

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 13]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
         }
     }
 
+    public int[][] calculateScore(Map<String, EmployeeProfile> employees){
+
+        Double[][] matrix = new Double[employees.values().size()][employees.values().size()]; //weight matrix
+        int[][] binaryMatrix = new int[employees.values().size()][employees.values().size()]; // binary matrix
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: make it boolean

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 14]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+                    binaryMatrix[i][j]=1;
+                else
+                    binaryMatrix[i][j]=0;
+            }
+
+        }
+        return binaryMatrix;
+    }
+
+    public void dfs(int[][] matrix, boolean[] visited, int node) {
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: can't name it dfs

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 15]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+        for (int i = 0; i < matrix[0].length; i++) {
+            if (matrix[node][i] == 1 && !visited[i]) {
+                visited[i] = true;
+                dfs(matrix, visited, i);
+            }
+        }
+    }
+
+    public int numOfSilos() {
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: calculateSilosCount

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 16]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+
+        for (int i = 0; i < size; i++) {
+            if (!visited[i]) {
+                count++;
+                visited[i] = true;
+                dfs(matrix, visited, i);
+            }
+        }
+        if(count == 0)
+            count++;
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: Don't need this

File: services/rule_engine/src/main/java/agents/CultureMaster.java [hunk 17]
```diff
@@ -79,4 +79,97 @@ public void sendEmployeesToAva() {
...
+        return count;
+    }
+
+    public String calculateRisk(int silosNum,int employeeNum){
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/agents/CultureMaster.java]: Risk of what

File: services/rule_engine/src/main/java/signals/NumOfSilosSignal.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+import com.mindsmiths.sdk.core.api.Signal;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class NumOfSilosSignal extends Signal {
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/signals/NumOfSilosSignal.java:12]: SilosCountSignal

Comment (mjuraj) [reply services/rule_engine/src/main/java/signals/NumOfSilosSignal.java:12]: Delete this signal

File: services/rule_engine/src/main/java/signals/SilosInfoSignal.java [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@Data
+@AllArgsConstructor
+public class SilosInfoSignal extends Signal {
+    private int numOfSilos;
```
Comment (mjuraj) [inline services/rule_engine/src/main/java/signals/SilosInfoSignal.java]: Send risk as well

File: services/rule_engine/src/main/resources/rules/Ava/Statistics.drl [hunk 20]
```diff
@@ -26,5 +30,25 @@ rule "Send stats email"
...
+    when
+        signal: SilosInfoSignal() from entry-point "signals"
+        agent: Ava()
+    then
+        modify(agent) {setNumOfSilos(signal.getNumOfSilos())};
+        delete(signal);
+end
+
+
+rule "Connect stats screens with email"
```
Comment (mjuraj) [inline services/rule_engine/src/main/resources/rules/Ava/Statistics.drl]: Show stats screens

**Review by mjuraj — CHANGES_REQUESTED** [review]

**Pull request closed without merging**